conwayste/src/ui/layer.rs

```diff
@@ -31,137 +32,245 @@ use super::{
+
+        if let Some(focused_slot) = self.focused_ids.last_mut() {
+            *focused_slot = Some(widget_id);
         }
 
-        let error_msg = format!("ERROR in Layer::enter_focus() =>  Widget (ID: {:?}) not found in layer (ID: {:?})",
-            id,
-            self.id
-        );
-        Err(Box::new(UIError::WidgetNotFound{reason: error_msg}))
+        Ok(())
     }
-}
 
-impl Widget for Layer {
-    fn id(&self) -> WidgetID {
-        self.id
-    }
+    /// Clears the focus of the highest layer
+    pub fn exit_focus(&mut self) {
+        let widget_id = *self.focused_ids.last().unwrap();
+
+        if let Some(widget_id) = widget_id {
+            if let Ok(dyn_widget) = self.get_widget_mut(widget_id) {
+                if let Some(tf_widget) = downcast_widget_mut!( dyn_widget, TextField) {
+                    tf_widget.exit_focus();
+                }
+            }
+        }
 
-    fn size(&self) -> Rect {
-        Rect::zero()
+        if let Some(widget_id) = self.focused_ids.last_mut() {
+            *widget_id = None;
+        }
     }
 
-    fn translate(&mut self, _dest: Vector2<f32>) {}
+    pub fn on_hover(&mut self, point: &Point2<f32>) {
+        let highest_z_index = self.peek_z_index();
 
-    fn on_hover(&mut self, point: &Point2<f32>) {
-        for w in self.widgets.iter_mut() {
-            w.on_hover(point);
+        for widget in self.widget_list.iter_mut()
+            .filter(|widget| widget.z_index() == highest_z_index)
+        {
+            widget.on_hover(point);
         }
     }
 
-    fn on_click(&mut self, point: &Point2<f32>) -> Option<(WidgetID, UIAction)> {
-        for w in self.widgets.iter_mut() {
-            let ui_action = w.on_click(point);
+    pub fn on_click(&mut self, point: &Point2<f32>) -> Option<(WidgetID, UIAction)> {
+        let highest_z_index = self.peek_z_index();
+
+        for widget in self.widget_list.iter_mut()
+            .filter(|widget| widget.z_index() == highest_z_index)
+        {
+            let ui_action = widget.on_click(point);
             if ui_action.is_some() {
                 return ui_action;
             }
         }
         None
     }
 
-    fn on_drag(&mut self, original_pos: &Point2<f32>, current_pos: &Point2<f32>) {
-        for w in self.widgets.iter_mut() {
-            w.on_drag(original_pos, current_pos);
+    pub fn on_drag(&mut self, original_pos: &Point2<f32>, current_pos: &Point2<f32>) {
+        let highest_z_index = self.peek_z_index();
+
+        for widget in self.widget_list.iter_mut()
+            .filter(|widget| widget.z_index() == highest_z_index)
+        {
+            widget.on_drag(original_pos, current_pos);
         }
     }
 
-    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
-        if self.with_transparency {
+    pub fn draw(&mut self, ctx: &mut Context) -> UIResult<()> {
+        let highest_z_index = self.peek_z_index();
+
+        if self.with_transparency && highest_z_index != 0 {
```

Comment:
Please explain why this test is the way it is.

Reply:
Found a behavioral bug with it due to your advise for clarifying the `Layering` comment. Bravo!